File: backend/app/schemas/workout.py
from datetime import date as date_

from pydantic import BaseModel, ConfigDict, Field


class WorkoutSetBase(BaseModel):
    exercise_id: int
    weight: float = Field(ge=0)
    reps: int = Field(ge=0)
    order: int = Field(ge=0)
    # Nulo = set suelto. Mismo entero = mismo bloque de super-serie dentro de
    # este workout (solo comparable dentro del mismo workout). Validado en el
    # router: un valor usado por menos de 2 exercise_id distintos es un 400.
    superset_group: int | None = None


class WorkoutSetCreate(WorkoutSetBase):
    pass


class WorkoutSet(WorkoutSetBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    workout_id: int


class WorkoutBase(BaseModel):
    date: date_
    routine_id: int | None = None


class WorkoutCreate(WorkoutBase):
    sets: list[WorkoutSetCreate] = []


class Workout(WorkoutBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sets: list[WorkoutSet] = []

File: mobile/src/types/workout.ts
// superset_group: null = serie suelta; mismo entero = misma super-serie.
// El ámbito de comparación es solo dentro de este workout, no es un id global.
export type WorkoutSetCreate = {
  exercise_id: number;
  weight: number;
  reps: number;
  order: number;
  superset_group: number | null;
};

export type WorkoutSet = WorkoutSetCreate & {
  id: number;
  workout_id: number;
};

export type WorkoutCreate = {
  date: string;
  routine_id: number | null;
  sets: WorkoutSetCreate[];
};

export type Workout = {
  id: number;
  date: string;
  routine_id: number | null;
  sets: WorkoutSet[];
};
